linode/instance/resource.go

```diff
@@ -688,25 +688,110 @@ func updateResource(ctx context.Context, d *schema.ResourceData, meta interface{
 		bootConfig = updatedConfigs[0].ID
 	}
 
+	booted := d.Get("booted").(bool)
+	bootedNull := d.GetRawConfig().GetAttr("booted").IsNull()
+
 	if d.HasChange("interface") {
 		interfaces := d.Get("interface").([]interface{})
 
+		powerOffRequired := false
+
 		expandedInterfaces := make([]linodego.InstanceConfigInterfaceCreateOptions, len(interfaces))
+		config, err := client.GetInstanceConfig(ctx, id, bootConfig)
+		if err != nil {
+			return diag.Errorf("failed to get the boot config: %s", err)
+		}
 
 		for i, ni := range interfaces {
 			expandedInterfaces[i] = helper.ExpandConfigInterface(ni.(map[string]interface{}))
+
+			newInterface := expandedInterfaces[i]
+			var oldInterface *linodego.InstanceConfigInterface
+			if len(config.Interfaces) > i {
+				oldInterface = &config.Interfaces[i]
+			}
+
+			vpcInterfaceInvolved := newInterface.Purpose == linodego.InterfacePurposeVPC ||
+				(oldInterface != nil && oldInterface.Purpose == linodego.InterfacePurposeVPC)
+
+			interfacePurposeChanged := oldInterface == nil || oldInterface.Purpose != newInterface.Purpose
+
+			powerOffRequired = vpcInterfaceInvolved && interfacePurposeChanged
 		}
 
 		tflog.Debug(ctx, "Updating instance config for interface changes", map[string]any{
 			"config_id": bootConfig,
 		})
 
-		if _, err := client.UpdateInstanceConfig(ctx, instance.ID, bootConfig, linodego.InstanceConfigUpdateOptions{
-			Interfaces: expandedInterfaces,
-		}); err != nil {
+		instance, err := client.GetInstance(ctx, id)
+		if err != nil {
+			return diag.Errorf("Error fetching data about the current linode: %s", err)
+		}
+
+		if powerOffRequired && meta.(*helper.ProviderMeta).Config.SkipImplicitReboots {
+			return diag.Errorf(
+				"add, remove, and reorder a Linode VPC interface requires implicit " +
+					"reboot of the Linode, please consider set 'skip_implicit_reboots' " +
+					"to true in the Linode provider config.",
+			)
+		}
+
+		if powerOffRequired {
+			if instance.Status != linodego.InstanceOffline {
+				if err := client.ShutdownInstance(ctx, id); err != nil {
+					return diag.Errorf(
+						"failed to shutdown instance: %s", err,
+					)
+				}
+			}
```

Comment:
I don't think this will handle transient states (e.g. shutting down), but I think the `applyBootStatus` helper in `instanceconfig` should cover those cases: https://github.com/linode/terraform-provider-linode/blob/dev/linode/instanceconfig/helper.go#L180

Reply:
I am trying to re-use some logic in `handleBootedUpdate` (by wrapping them up in a function) to handle the cases of transient state. Do you think that looks good?

Reply:
Looks good to me ^^